src/types/objects.rs

```diff
@@ -0,0 +1,85 @@
+//! Implementations for the types describing DIF object files.
+
+use super::{
+    AllObjectCandidates, Arch, CodeId, CompleteObjectInfo, DebugId, ObjectCandidate,
+    ObjectFeatures, ObjectFileStatus, ObjectUseInfo, RawObjectInfo,
+};
+use crate::sources::{SourceId, SourceLocation};
+
+impl Default for ObjectUseInfo {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+impl ObjectUseInfo {
+    pub fn is_none(&self) -> bool {
+        matches!(self, Self::None)
+    }
+}
+
+impl From<Vec<ObjectCandidate>> for AllObjectCandidates {
+    fn from(mut source: Vec<ObjectCandidate>) -> Self {
+        source.sort_by_key(|candidate| (candidate.source.clone(), candidate.location.clone()));
```

Comment:
The docs say:

> For expensive key functions (e.g. functions that are not simple property accesses or basic operations), sort_by_cached_key is likely to be significantly faster, as it does not recompute element keys.

Since at least `location.clone()` allocates a new string, you should try to optimize this.

Reply:
Oh yes, didn't spot the cached version.

This cloning is annoying, I didn't know what to do better about this.

Reply:
hm, why can’t you just use `(&candidate.source, &candidate.location)`? oh yes, because sort is `mut`, or does it basically take care of that internally?

Reply:
I think it's the tuple that makes this impossible.